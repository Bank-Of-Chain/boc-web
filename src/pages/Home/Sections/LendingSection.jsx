import React, { useEffect, useState } from "react"
import classNames from "classnames"
import { makeStyles } from "@material-ui/core/styles"

// === Components === //
import GridContainer from "../../../components/Grid/GridContainer"
import GridItem from "../../../components/Grid/GridItem"
import CircularProgress from "@material-ui/core/CircularProgress"
import useMediaQuery from '@material-ui/core/useMediaQuery'
import Tooltip from "@material-ui/core/Tooltip"

// === Utils === //
import get from "lodash/get"
import map from "lodash/map"
import maxBy from "lodash/maxBy"
import isNaN from "lodash/isNaN"
import filter from "lodash/filter"
import sortBy from "lodash/sortBy"
// import { calVaultAPY } from "./../../../helpers/apy"
import { toFixed } from "./../../../helpers/number-format"
import { getDefiRate } from "./../../../services/api-service"
// import { getETHLast30DaysVaultData } from "./../../../services/subgraph-service"

// === Styles === //
import styles, { smStyle } from "./lendingStyle"

const useStyles = makeStyles(styles)
const useSmStyles = makeStyles(smStyle)

const bocTitle = "BOC"

const array = ["Compound", "Aave", "Coinbase", "BlockFi", "Nexo", "Celsius", "YearnFinance", "Gemini", "Bitfinex"]
const apyType = {
  BlockFi: "Fixed Rate",
  Celsius: "Fixed Rate",
  Nexo: "Fixed Rate",
  Gemini: "Fixed Rate",
  Coinbase: "Fixed Rate",
  YearnFinance: "Current Rate",
  Compound: "Current Rate",
  Aave: "Current Rate",
  BOC: "Current Rate",
  Bitfinex: "Fixed Rate",
}
export default function LendingSection () {
  const isLayoutSm = useMediaQuery('(max-width: 960px)')
  const smClasses = useSmStyles()
  const classes = useStyles({ classes: isLayoutSm ? smClasses : {} })
  const [loading, setLoading] = useState(false)
  const [data, setData] = useState([])

  useEffect(() => {
    setLoading(true)
    Promise.all([
      // getETHLast30DaysVaultData().then(a => {
      //   return {
      //     title: bocTitle,
      //     imagePath: "/logo.png",
      //     percent: (100 * calVaultAPY(a)).toFixed(2),
      //     text: get(apyType, bocTitle, ""),
      //   }
      // }),
      getDefiRate().catch(() =>
        Promise.resolve({
          data: {},
          svg: {},
        }),
      ),
    ])
      // .then(([obj, resp]) => {
      .then(([resp]) => {
        const { data, svg } = resp
        return [
          // obj,
          ...map(array, i => {
            return {
              title: i === "YearnFinance" ? "Yearn" : i,
              imagePath: svg[i],
              percent: parseFloat(data[i]),
              text: get(apyType, i, ""),
            }
          }),
        ]
      })
      .then(setData)
      .finally(() => {
        setTimeout(() => {
          setLoading(false)
        }, 1000)
      })
  }, [])
  const maxPercentItem = maxBy(data, "percent")
  const displayMaxValue = 5 * Math.ceil(maxPercentItem?.percent / 5)
  return (
    <div className={classes.section}>
      <h2 className={classes.title}>Crypto Lending Interest Rates</h2>
      <div className={classNames(classes.label)}>
        <GridContainer>
          <GridItem>
            <div className={classNames(classes.box)}></div>
            <span style={{ verticalAlign: 'top', fontSize:'0.75rem' }}>Current Rate</span>
          </GridItem>
          <GridItem>
            <div className={classNames(classes.box1)}></div>
            <span style={{ verticalAlign: 'top', fontSize:'0.75rem' }}>Fixed  Rate</span>
          </GridItem>
        </GridContainer>
      </div>
      <div style={{ padding: "4.5rem 0" }}>
        {loading ? (
          <GridContainer style={{ margin: "0 auto" }} justify='center'>
            <CircularProgress />
          </GridContainer>
        ) : (
          <GridContainer style={{ margin: "0 auto" }} justify='center'>
            {map(
              filter(
                sortBy(data, o => {
                  if (o.title === bocTitle) {
                    return -10000
                  }
                  return -1 * o.percent
                }),
                o => !isNaN(o.percent),
              ),
              (item, i) => {
                const { title, imagePath, percent, text = "" } = item
                const nextPercent = percent / displayMaxValue
                const percentText = `${toFixed(nextPercent.toString(), 1e-2, 2)}%`
                return (
                  <GridItem className={classNames(classes.item)} key={`${i}`} xs={1} sm={1} md={1}>
                    <GridContainer className={classes.body}>
                      <GridItem className={classes.header} style={i === 0 ? { borderLeft: 0 } : {}}>
                        <Tooltip title={text}>
                          <div
                            className={classNames(classes.bar, text === 'Current Rate' && classes.fixed, title === bocTitle && classes.checked)}
                            style={{ height: percentText }}
                          >
                            <p>
                              {percent}%
                            </p>
                          </div>
                        </Tooltip>
                      </GridItem>
                      <GridItem className={classes.footer}>
                        <img title={title} src={imagePath} alt={title} />
                        <p>{title}</p>
                      </GridItem>
                    </GridContainer>
                  </GridItem>
                )
              },
            )}
          </GridContainer>
        )}
      </div>
    </div>
  )
}
