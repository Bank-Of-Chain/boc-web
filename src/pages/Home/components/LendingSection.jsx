import React, { useEffect, useState } from 'react'
import classNames from 'classnames'
import { makeStyles } from '@material-ui/core/styles'

// === Components === //
import GridContainer from '@/components/Grid/GridContainer'
import GridItem from '@/components/Grid/GridItem'
import CircularProgress from '@material-ui/core/CircularProgress'
import useMediaQuery from '@material-ui/core/useMediaQuery'
import Tooltip from '@material-ui/core/Tooltip'

// === Utils === //
import get from 'lodash/get'
import map from 'lodash/map'
import maxBy from 'lodash/maxBy'
import isNaN from 'lodash/isNaN'
import filter from 'lodash/filter'
import sortBy from 'lodash/sortBy'
import { toFixed } from '@/helpers/number-format'
import { getDefiRate, getAPY } from '@/services/api-service'

// === Styles === //
import styles, { smStyle } from './lendingStyle'

const useStyles = makeStyles(styles)
const useSmStyles = makeStyles(smStyle)

const bocTitle = 'BoC'

const array = ['Compound', 'Aave', 'Coinbase', 'YearnFinance', 'Gemini', 'Bitfinex', 'Staking']
const apyType = {
  BlockFi: 'Fixed Rate',
  Nexo: 'Fixed Rate',
  Gemini: 'Fixed Rate',
  Coinbase: 'Fixed Rate',
  YearnFinance: 'Current Rate',
  Compound: 'Current Rate',
  Aave: 'Current Rate',
  BoC: 'Current Rate',
  Bitfinex: 'Fixed Rate'
}

export default function LendingSection() {
  const isLayoutSm = useMediaQuery('(max-width: 960px)')
  const smClasses = useSmStyles()
  const classes = useStyles({ classes: isLayoutSm ? smClasses : {} })
  const [loading, setLoading] = useState(false)
  const [data, setData] = useState([])
  const [ethiData, setEthiData] = useState([])

  useEffect(() => {
    setLoading(true)
    const chainId = '1'
    Promise.all([
      getAPY({ chainId, tokenType: 'USDi' }).then((data = 0) => {
        return {
          title: bocTitle,
          imagePath: '/logo.png',
          percent: parseFloat(data).toFixed(2),
          text: get(apyType, bocTitle, '')
        }
      }),

      getDefiRate(chainId, 'USDi').catch(() =>
        Promise.resolve({
          data: {},
          svg: {}
        })
      ),
      getAPY({ chainId, tokenType: 'ETHi' }).then((data = 0) => {
        return {
          title: bocTitle,
          imagePath: '/logo.png',
          percent: parseFloat(data).toFixed(2),
          text: get(apyType, bocTitle, '')
        }
      }),
      getDefiRate(chainId, 'ETHi').catch(() =>
        Promise.resolve({
          data: {},
          svg: {}
        })
      )
    ])
      .then(([obj, resp, ethiObj, ethiResp]) => {
        const { data, svg } = resp
        setData([
          obj,
          ...map(array, i => {
            return {
              title: i === 'YearnFinance' ? 'Yearn' : i,
              imagePath: svg[i],
              percent: parseFloat(data[i]),
              text: get(apyType, i, '')
            }
          })
        ])
        setEthiData([
          ethiObj,
          ...map(array, i => {
            return {
              title: i === 'YearnFinance' ? 'Yearn' : i,
              imagePath: ethiResp.svg[i],
              percent: parseFloat(ethiResp.data[i]),
              text: get(apyType, i, '')
            }
          })
        ])
      })
      .finally(() => {
        setTimeout(() => {
          setLoading(false)
        }, 1000)
      })
  }, [])

  const renderJsx = (array = [], displayMaxValue) => {
    return (
      <GridContainer className={classes.block}>
        {map(
          filter(
            sortBy(array, o => {
              if (o.title === bocTitle) {
                return -10000
              }
              return -1 * o.percent
            }),
            o => !isNaN(o.percent)
          ),
          (item, i) => {
            const { title, imagePath, percent, text = '' } = item
            const nextPercent = percent / displayMaxValue
            const percentText = `${toFixed(nextPercent.toString(), 1e-2, 2)}%`
            return (
              <GridItem className={classNames(classes.item)} key={`${i}`} xs={12} sm={12} md={12}>
                <GridContainer className={classes.body}>
                  <GridItem className={classes.footer} xs={3} sm={3} md={3}>
                    <img
                      title={title}
                      src={imagePath}
                      alt={title}
                      onError={({ currentTarget }) => {
                        currentTarget.onerror = null // prevents looping
                        currentTarget.src = '/default.png'
                      }}
                    />
                    <p>{title}</p>
                  </GridItem>
                  <GridItem className={classes.header} style={i === 0 ? { borderLeft: 0 } : {}} xs={9} sm={9} md={9}>
                    <Tooltip title={text}>
                      <div
                        className={classNames(classes.bar, text === 'Fixed Rate' && classes.fixed, title === bocTitle && classes.checked)}
                        // style={{ width: `calc(${percentText} - 4rem)` }}
                        style={{ width: percentText }}
                      >
                        <p className={classes.percent}>{percent}%</p>
                      </div>
                    </Tooltip>
                  </GridItem>
                </GridContainer>
              </GridItem>
            )
          }
        )}
      </GridContainer>
    )
  }

  const maxPercentItem = maxBy([...data, ...ethiData], 'percent')
  // left 25% percent width for text render
  const displayMaxValue = 1.25 * maxPercentItem?.percent
  return (
    <div className={classes.section}>
      <h2 className={classes.title}>Crypto Lending Interest Rates</h2>

      <div className={classes.chart}>
        <div className={classes.chartInner}>
          {loading ? (
            <GridContainer style={{ margin: '0 auto', padding: '10rem' }} justify="center">
              <CircularProgress />
            </GridContainer>
          ) : (
            <GridContainer className={classes.block} style={{ margin: '0 auto' }} justify="center">
              <GridItem sm={12} xs={6} md={6} lg={6} style={{ backgroundColor: 'rgba(0,0,0,0.1)' }}>
                <p className={classes.colorful}>USDi</p>
                {renderJsx(data, displayMaxValue)}
              </GridItem>
              <GridItem sm={12} xs={6} md={6} lg={6}>
                <p className={classes.colorful}>ETHi</p>
                {renderJsx(ethiData, displayMaxValue)}
              </GridItem>
            </GridContainer>
          )}
        </div>
      </div>
      {!loading && (
        <div className={classNames(classes.label)}>
          <GridContainer>
            <GridItem style={{ lineHeight: '2rem' }}>
              <div className={classNames(classes.box)}></div>
              <span>Current Rate</span>
            </GridItem>
            <GridItem>
              <div className={classNames(classes.box1)}></div>
              <span>Fixed Rate</span>
            </GridItem>
          </GridContainer>
        </div>
      )}
    </div>
  )
}
