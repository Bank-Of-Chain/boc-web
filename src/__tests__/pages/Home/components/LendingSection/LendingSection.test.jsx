import React from 'react'
import { render } from '@testing-library/react'
import LendingSection from '@/pages/Home/components/LendingSection'

test('LendingSection Component Render', () => {
  const { asFragment } = render(<LendingSection />)
  expect(asFragment()).toMatchSnapshot()
})
