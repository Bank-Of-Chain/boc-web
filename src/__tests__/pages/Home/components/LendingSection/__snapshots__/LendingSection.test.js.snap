// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LendingSection Component Render 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-section-4"
  >
    <h2
      class="makeStyles-title-6"
    >
      Crypto Lending Interest Rates.
    </h2>
    <div
      class="makeStyles-label-13"
    >
      <div
        class="MuiGrid-root makeStyles-grid-18  MuiGrid-container"
      >
        <div
          class="MuiGrid-root makeStyles-grid-19  MuiGrid-item"
          style="line-height: 2rem;"
        >
          <div
            class="makeStyles-box-14"
          />
          <span
            style="font-size: 0.75rem;"
          >
            Current Rate
          </span>
        </div>
        <div
          class="MuiGrid-root makeStyles-grid-19  MuiGrid-item"
        >
          <div
            class="makeStyles-box1-15"
          />
          <span
            style="font-size: 0.75rem;"
          >
            Fixed Rate
          </span>
        </div>
      </div>
    </div>
    <div
      class="makeStyles-chart-16"
    >
      <div
        class="makeStyles-chartInner-17"
      >
        <div
          class="MuiGrid-root makeStyles-grid-18  MuiGrid-container MuiGrid-justify-xs-center"
          style="margin: 0px auto;"
        >
          <div
            class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
